### PR TITLE
T196 fix additional frame

### PR DESCRIPTION
**Pull request by Funth0mas**
Extend `--additionalFrameAttributes` to support any attribute name. Add this feature to json export. Replace `eval()` by `getattr()`.

### CODE REVIEW

**Comment by altendky** [comment]
Yeah, `eval()` is...  ugly.  But, I'm not a huge fan of making dict keys available as attributes either.  The cli effect could be achieved without that as well.  I can't say I have a really strong argument, it mostly just seems less 'magical'.  (and yes, I have worked with `__getattr__` etc, so I don't mean magical like I am not familiar with it)

```python
getattr(signal, item, signal.attributes.get(item, ""))
```

Or just add a getter.

```python
NOTHING = object()
def get(self, item, default=NOTHING):
    item = getattr(self, item, self.attributes.get(item, default))

    if item is NOTHING:
        raise SomeException()

    return item
```
```python
signal.get(item, "")
```

**Comment by Funth0mas** [comment]
Totally agree, your solution is better. I'll apply it.

There is still question if it make sense to have some "additional attributes" as instance variables and some in the attributes dictionary. It would make sense to have them all on a single place. But I didn't want to touch the existing code, because...  I'm just a passerby.

**Comment by altendky** [comment]
If I remember a conversation with @ebroecker correctly, it would be nice to revisit these 'attributes' and try to figure out which are common and how to nestle the rest together happily.  Basically, there is a dbc bias and with the extra knowledge of other formats we could probably do a better job picking regular attributes to have.

Thanks for seeing the mess and doing something about at least part of it.  :]  We appreciate the help.

**Comment by ebroecker** [comment]
@Funth0mas 
having all attributes in one place is not that easy...
My intention was to have the "common" attributes direct in the frame (yes some are missing). 
The attributes[]-Dictionary is more or less a "dbc-only" place to save all dbc-attributes. The problem of dbc is, that you can define any attribute in dbc. 
Some attributes are not yet in the common list (#146)

File: src/canmatrix/canmatrix.py [hunk 1]
```diff
@@ -425,62 +436,75 @@ def pgn(self, value):
...
+
+    def attribute(self, name, db=None, default=None):
+        """Get any Frame attribute by its name.
+
+        :param str name: attribute name, can be mandatory (ex: id) or optional (customer) attribute.
+        :param CanMatrix db: Optional database parameter to get global default attribute value.
+        :param default: Default value if attribute doesn't exist.
+        :return any: Return the attribute value if found, else `default` or None
+        """
+        if hasattr(self, name):
```
Comment (altendky) [inline src/canmatrix/canmatrix.py]: Sorry to evolve my request here...  I forgot this was an `attrs` class and I think this will be a touch better.  It will avoid considering methods and such as attributes.

https://www.attrs.org/en/stable/api.html#attr.fields_dict

```python
if name in attr.fields_dict(type(self)):
```

I'll note that this is technically a breaking change given the it will block access via this method to `.attribute` items that shadow the names of regular attributes.  But, I don't think that bothers me.

File: src/canmatrix/canmatrix.py [hunk 2]
```diff
@@ -425,62 +436,75 @@ def pgn(self, value):
...
+    #    def cycleTime(self, value):
+    #        self._cycleTime = value
+    #        self.attributes["GenMsgCycleTime"] = value
+
+    #    @sendType.setter
+    #    def sendType(self, value):
+    #       self._sendType = value
+    #       self.attributes["GenMsgCycleTime"] = value
+
+    def attribute(self, name, db=None, default=None):
```
Comment (altendky) [inline src/canmatrix/canmatrix.py]: This is definitely a breaking change though it does seem better.  I think it's better for `canmatrix` to be tending towards improving so I'm comfortable with this but figured I'd highlight it.

File: src/canmatrix/canmatrix.py [hunk 3]
```diff
@@ -425,62 +436,75 @@ def pgn(self, value):
...
+            return self.attributes[name]
         elif db is not None:
-            if attributeName in db.frameDefines:
-                define = db.frameDefines[attributeName]
+            if name in db.frameDefines:
+                define = db.frameDefines[name]
                 return define.defaultValue
         else:
-            return None
+            return default
```
Comment (altendky) [inline src/canmatrix/canmatrix.py]: +1

File: src/canmatrix/cmjson.py [hunk 4]
```diff
@@ -101,21 +101,21 @@ def dump(db, f, **options):
                               "signals": signals}
             frame_attributes = {}
             for frame_info in additionalFrameColums:  # Look for additional Frame Attributes
-                if hasattr(frame, frame_info):
-                    frame_attributes[frame_info] = getattr(frame, frame_info)
+                if frame.attribute(frame_info):  # does the attribute exist?
```
Comment (altendky) [inline src/canmatrix/cmjson.py]: This doesn't check it's existence.  If it exists but is `None` or anything else falsey it will be treated as not present.  Here's the first thing that came to my mind, though it is a touch smelly to me.  Maybe you can think of something nicer.

```python
MISSING = object()
value = frame.attribute(name=frame_info, default=MISSING)
if value is not MISSING:
    frame_attributes[frame_info] = value
```

File: src/canmatrix/cmjson.py [hunk 5]
```diff
@@ -101,21 +101,21 @@ def dump(db, f, **options):
...
             if frame_attributes:  # only add Attributes if there are any
                 symbolic_frame["attributes"] = frame_attributes
             exportArray.append(symbolic_frame)
-    else:  # exportall
+    else:  # exportAll
         for frame in db.frames:
             frameattribs = {}
             for attribute in db.frameDefines:
-                frameattribs[attribute] = frame.attribute(db,attribute)
+                frameattribs[attribute] = frame.attribute(attribute, db)
```
Comment (altendky) [inline src/canmatrix/cmjson.py]: Let's specify by name to avoid this hazard in the future.  Looks like this applies to most of the rest of the diff below.

**Comment by Funth0mas** [comment]
I ran the autoformating on the whole file by accident. If it's problem, I'll revert it line per line. Sorry for the inconvenience. 
Mainly I removed the cryptic getattr and I propose to use the existing attribute() getter, but make the db parameter really optional. I added some docstrings (only in Frame) when trying to understand the code.

**Comment by altendky** [comment]
No big deal, I just tend to mention more rather than less.  Thanks for the help improving the code.

File: src/canmatrix/canmatrix.py [hunk 6]
```diff
@@ -370,45 +374,52 @@ def __str__(self):
...
-    transmitters (list of boardunits/ECU-names),
-    size (= DLC),
-    signals (list of signal-objects),
-    attributes (list of attributes),
-    receiver (list of boardunits/ECU-names),
-    extended (Extended Frame = 1),
-    comment
+    Contains one CAN Frame.
+
+    The Frame has  following mandatory attributes\n
```
[the remark this answers is not in the document]
Comment (Funth0mas) [reply src/canmatrix/canmatrix.py]: There must be either empty line or \n befor the list (ctrl+Q in PyCharm). Changed to empty line.

File: src/canmatrix/canmatrix.py [hunk 7]
```diff
@@ -105,49 +106,50 @@ class Signal(object):
             _multiplex ('Multiplexor' or Number of Multiplex)
     """
 
-    name = attr.ib(default = "")
-#    float_factory = attr.ib(default=defaultFloatFactory)
+    name = attr.ib(default="")
+    #    float_factory = attr.ib(default=defaultFloatFactory)
```
[the remark this answers is not in the document]
Comment (Funth0mas) [reply src/canmatrix/canmatrix.py]: (comic is that it was done automatically by PyCharm) ... clear, reverted, the same for all another unwanted changes. Comments, indentation and another codestyle issues can be fixed later in own PR.

File: src/canmatrix/canmatrix.py (hunk 2 above)
Comment (Funth0mas) [reply src/canmatrix/canmatrix.py]: If this change is accepted, I'd adopt the similar functions of another classes accordingly.

File: src/canmatrix/cmjson.py (hunk 4 above)
Comment (Funth0mas) [reply src/canmatrix/cmjson.py]: good point! As there is imho no reason to export None value, change to `is not None`

File: src/canmatrix/cmjson.py (hunk 5 above)
Comment (Funth0mas) [reply src/canmatrix/cmjson.py]: Uff, I realized the PyCharm has refactored also calls of another classes. Python has certain weaknesses :-(
Checked, reverted unwanted changes. For Frame use named attribute.

File: src/canmatrix/canmatrix.py [hunk 8]
```diff
@@ -780,7 +816,8 @@ def __init__(self, definition):
         def saveConvertStrToInt(inStr):
             out = int(defaultFloatFactory(inStr))
             if out != defaultFloatFactory(inStr):
-                logger.warning("Warning, integer was expected but got float: got: {0} using {1}\n".format(inStr, str(out)))
+                logger.warning(
+                    "Warning, integer was expected but got float: got: {0} using {1}\n".format(inStr, str(out)))
```
[the remark this answers is not in the document]
Comment (Funth0mas) [reply src/canmatrix/canmatrix.py]: unwanted change, revert.

**Comment by Funth0mas** [comment]
Ready for review again. Thanks!

File: src/canmatrix/canmatrix.py [hunk 9]
```diff
@@ -409,6 +416,7 @@ class Frame(object):
 
     @property
     def is_multiplexed(self):
+        """Frame is multiplexed if at least one of its signals is multiplexed."""
```
Comment (ebroecker) [inline src/canmatrix/canmatrix.py]: Frame is multiplexed it at least one of its signals is a multiplexer

File: src/canmatrix/canmatrix.py (hunk 2 above)
Comment (ebroecker) [reply src/canmatrix/canmatrix.py]: canmatrix is at 0. - version. I agree that improving is always better than staying compatible.  
I'd accept it!

File: src/canmatrix/canmatrix.py [hunk 10]
```diff
@@ -552,45 +584,54 @@ def globSignals(self, globStr):
...
         """
-        add attribute to attribute-list of frame; If Attribute already exits, modify value
+        Add the attribute with value to customer Frame attribute-list. If Attribute already exits, modify its value.
+        :param str attribute: Attribute name
+        :param any value: attribute value
         """
         try:
             self.attributes[attribute] = str(value)
-        except:
+        except:  # could it really fail?
```
Comment (ebroecker) [inline src/canmatrix/canmatrix.py]: I'm not sure why I added this try-catch. But I'm quite sure that there was a problem without it.
Maybe there is a problem with "str" if your are in python2 and have some issues with charsets...

File: src/canmatrix/cmcsv.py [hunk 11]
```diff
@@ -183,10 +183,7 @@ def dump(db, thefile, delimiter=',', **options):
 
         additionalFrameInfo = []
         for frameInfo in additionalFrameCollums:
-            try:
-                temp = eval("frame." + frameInfo)
-            except:
-                temp = ""
+            temp = frame.attribute(frameInfo, default="")
```
Comment (ebroecker) [inline src/canmatrix/cmcsv.py]: +++ this eliminates my dirty "eval"-hack.
Thanks

File: src/canmatrix/canmatrix.py (hunk 10 above)
Comment (altendky) [reply src/canmatrix/canmatrix.py]: None of this should be dealt with in this PR but:

1) it should catch the particular exception not just anything
2) i think users should be expected to pass the proper type
3) i think there are still various str-or-int-or-... around so maybe it would get caught up in cleaning that up.

File: src/canmatrix/cmcsv.py [hunk 12]
```diff
@@ -270,3 +261,7 @@ def dump(db, thefile, delimiter=',', **options):
 #        finalTableString = "\n".join(
 #            [row.toCSV(delimiter) for row in csvtable])
 #        print(finalTableString)
+    if sys.version_info > (3, 0):
+        # When TextIOWrapper is garbage collected, it closes the raw stream
+        # unless the raw stream is detached first
+        temp.detach()
```
Comment (altendky) [inline src/canmatrix/cmcsv.py]: Funny thing, I just ran into this over in a tqdm PR.  Sensible addition given the `if` above but the thing overall sort of smells to me.  `*hopes he wasn't the one that added it*`

Should probably really be in a finally clause though just in case there are exceptions.  Or a context manager, whatever you feel like.

Comment (Funth0mas) [reply src/canmatrix/cmcsv.py]: I simply copied it from another module because there were an error/warning after running the Unit Test informing about unreleased resource. To be honest I don't know why the `TextIOWrapper` must be present. Either let it be for further cleanup (I can create a ticket) or I need more time to investigate. The first solution is currently more convenient for me :-)

File: src/canmatrix/canmatrix.py (hunk 2 above)
Comment (Funth0mas) [reply src/canmatrix/canmatrix.py]: I'll refactor the `attribute()` of another classes to have the same signature in another PR (when I don't fall asleep in the train next time)

**Comment by Funth0mas** [comment]
The failed appveyor isn't my fault  :-(

File: src/canmatrix/cmcsv.py (hunk 12 above)
Comment (altendky) [reply src/canmatrix/cmcsv.py]: Yeah, all I mean for this is to make a try/finally to be sure the detach actually gets run even in case of exceptions.

**Comment by Funth0mas** [comment]
Hi, I merged the current origin/development and now it seems I did more changes than I really did, when looking to "changed files". I'm used to do this with bitbucket but it has different result. Did I something wrong?

**Comment by altendky** [comment]
@Funth0mas are you referring to 3bdafa9?  I generally do a `git merge --no-ff development` and get a merge commit.  I'm not really sure quite why it was diffing the way it was...  but I think I just fixed it.  Fwiw, I did a `git rebase -i HEAD~3` and deleted the squash commit then did a regular `git merge development --no-ff` and a force push to your branch (you let us have write access to just that branch when you submitted the PR).  Yes, I did also push a backup branch to my repo first.  :]

Anyways, if this looks right to you I'm ready to merge.

**Comment by Funth0mas** [comment]
@altendky funny thing is that I did almost the same except of using more complicated `git reset` and force pushed two minutes before you did :-) At least I understand the git a little bit more now. 
I 'inspected' the diff once again and yes, it's the state I wanted to have. Thank you for your attention.

**Comment by altendky** [comment]
Thanks for all the code and your responsiveness.

Cheers,
-kyle

**Comment by Funth0mas** [comment]
My pleasure. BTW you're from Germany?
Regards

**Comment by ebroecker** [comment]
No, only I am from Germany, altendkey Not...
